lib/src/spotify_base.dart

```diff
@@ -96,31 +87,50 @@ abstract class SpotifyApiBase {
   }
 
   Future<String> _getImpl(String url, Map<String, String> headers) async {
-    final response = await (await _client).get(url, headers: headers);
-    return handleErrors(response);
+    return await _requestWrapper(
+        () async => await (await _client).get(url, headers: headers));
   }
 
   Future<String> _postImpl(
       String url, Map<String, String> headers, dynamic body) async {
-    var response =
-        await (await _client).post(url, headers: headers, body: body);
-    return handleErrors(response);
+    return await _requestWrapper(() async =>
+        await (await _client).post(url, headers: headers, body: body));
   }
 
   Future<String> _deleteImpl(
       String url, Map<String, String> headers, body) async {
-    final request = http.Request('DELETE', Uri.parse(url));
-    request.headers.addAll(headers);
-    request.body = body;
-    final response =
-        await http.Response.fromStream(await (await _client).send(request));
-    return handleErrors(response);
+    return await _requestWrapper(() async {
+      final request = http.Request('DELETE', Uri.parse(url));
+      request.headers.addAll(headers);
+      request.body = body;
+      return await http.Response.fromStream(
+          await (await _client).send(request));
+    });
   }
 
   Future<String> _putImpl(
       String url, Map<String, String> headers, dynamic body) async {
-    var response = await (await _client).put(url, headers: headers, body: body);
-    return handleErrors(response);
+    return await _requestWrapper(() async =>
+        await (await _client).put(url, headers: headers, body: body));
+  }
+
+  Future<String> _requestWrapper(Future<http.Response> Function() request,
+      {retryLimit = 5}) async {
+    for (var i = 0; i < retryLimit; i++) {
+      while (_shouldWait){
+        await Future.delayed(Duration(milliseconds: 500));
+      }
+      try {
+        return handleErrors(await request());
+      } on ApiRateException catch (ex) {
+        if (i == retryLimit - 1) rethrow;
```

Comment:
Could you please add test cases for this exception? For when the exception actually gets thrown after a while and maybe if possible one for when it just retries for some time and finally succeeds?

Reply:
[There is already a way to set the mock http error](https://github.com/rinukkusu/spotify-dart/blob/master/test/spotify_test.dart#L46) - maybe you need to extend it with the wait time or something along the lines 😄

Reply:
Added the tests.
It turned out that the existing error tests didn't work at all, the setter set the mock error property of `SpotifyApiMock`, but the request checked `MockClient`'s error (that was always null) so the error tests generated successful responses and the expect statements were in the catch block.
After I fixed this bug, after the first error was set, the rest of the test failed either. So instead of setting the error back to `null` after each test, I changed the mock error property to be an `Iterator` which helped to implement the rate tests too.

Reply:
Looks really great - thanks for also fixing the broken tests!!